performance/single-node/builtin_atomic_tests.c

```diff
@@ -0,0 +1,323 @@
+
+
+#define FETCH_ADD \
+dbg_printf("atomic_var = %d\n", atomic_fetch_add(&atomic_var, 1));
+
+#define FETCH_ADD_EXPLICIT \
+dbg_printf("atomic_var = %d\n", atomic_fetch_add_explicit(&atomic_var, 1, MEMORY_ORDER));
+
+
+#define FETCH_SUB \
+dbg_printf("atomic_var = %d\n", atomic_fetch_sub(&atomic_var, 1));
+
+#define FETCH_SUB_EXPLICIT \
+dbg_printf("atomic_var = %d\n", atomic_fetch_sub_explicit(&atomic_var, 1, MEMORY_ORDER));
+
+
+#define FETCH_OR \
+atomic_fetch_or(&atomic_var, ++val);
+
+#define FETCH_OR_EXPLICIT \
+atomic_fetch_or_explicit(&atomic_var, ++val, MEMORY_ORDER);
+
+
+#define FETCH_XOR \
+atomic_fetch_xor(&atomic_var, ++val);
+
+#define FETCH_XOR_EXPLICIT \
+atomic_fetch_xor_explicit(&atomic_var, ++val, MEMORY_ORDER);
+
+
+#define FETCH_AND \
+atomic_fetch_and(&atomic_var, ++val);
+
+#define FETCH_AND_EXPLICIT \
+atomic_fetch_and_explicit(&atomic_var, ++val, MEMORY_ORDER);
+
+
+#define FLAG_TEST_AND_SET \
+atomic_flag_test_and_set(&atomic_var);
+
+#define FLAG_TEST_AND_SET_EXPLICIT \
+atomic_flag_test_and_set_explicit(&atomic_var, MEMORY_ORDER);
+
+
+#define FLAG_CLEAR \
+atomic_flag_clear(&atomic_var);
+
+#define FLAG_CLEAR_EXPLICIT \
+atomic_flag_clear_explicit(&atomic_var, memory_order_relaxed);
+/*******************************************************************************
+ * End wrappers
+ ******************************************************************************/
+
+/*******************************************************************************
+ * Begin test functions
+ ******************************************************************************/
+#define DO_TEST_LOOP(FN, ITERS)			\
+	do {					\
+		assert((ITERS));		\
+		volatile uint64_t i = (ITERS);	\
+		while (i--) { FN }		\
+	} while (0)
+
+#define RUN_TIMED_TEST(NAME, FN, ITERS)					\
+	do {								\
+		dbg_printf("Starting %s...\n", NAME);			\
+		start_time = get_time_nsec();				\
+		DO_TEST_LOOP(FN, iters);				\
+		stop_time = get_time_nsec();				\
+		printf("atomic_var = %d\n", atomic_fetch_add(&atomic_var, 0)); \
+		printf("dummy_val = %d\n", dummy_val);				\
+		ct_add_line(info, ":%s: %-lu\n", NAME, stop_time - start_time, FIELD_WIDTH); \
+	} while (0)
+/*******************************************************************************
+ * End test functions
+ ******************************************************************************/
+
+/*******************************************************************************
+ * Run tests
+ ******************************************************************************/
+int main(int argc, char **argv)
+{
+	char *test_name   = "---gnu_gcc_atomics---";
+	char *csv_header  = ":functionName: runtime(ns)";
+	char *csv_path	  = "./builtin_atomic_tests.csv"; // Touch to create
+	ct_info_t *info;
+	uint64_t iters, start_time, stop_time; // ~4*10^6 iterations
+
+	info = ct_init_info(test_name, csv_path);
+
+	/* Attempt to prevent compiler from unrolling the loop */
```

Comment:
I just want the loop to be as-is (not unrolled), in this case, it probably doesn't matter one way or the other.